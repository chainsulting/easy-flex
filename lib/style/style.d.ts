import { FC, HTMLAttributes } from 'react';
import { IColor, ICssLineHeight, IFontSize, IFontWeight, ILineHeight, ISize, IStyleElement } from '../types';
export interface IStyleProps extends HTMLAttributes<HTMLSpanElement> {
    /** Component's background color. */
    backgroundColor?: IColor;
    /** Component's color. */
    color?: IColor;
    /** Component's html tag. */
    element?: IStyleElement;
    /** Component's font size. */
    fontSize?: IFontSize | ISize;
    /** Component's font weight. */
    fontWeight?: IFontWeight | number;
    /** If true, the text style is set to italic. */
    italic?: boolean;
    lineHeight?: ICssLineHeight | ILineHeight;
}
export declare const Style: FC<IStyleProps>;
