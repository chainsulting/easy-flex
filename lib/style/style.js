"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Style = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("react");
var styled_components_1 = __importDefault(require("styled-components"));
var utils_1 = require("../utils");
var StyledStyle = styled_components_1.default.span(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n\tdisplay: flex;\n\tbox-sizing: border-box;\n\tbackground-color: ", ";\n\tcolor: ", ";\n\tfont-size: ", ";\n\tfont-weight: ", ";\n\tfont-style: ", ";\n"], ["\n\tdisplay: flex;\n\tbox-sizing: border-box;\n\tbackground-color: ", ";\n\tcolor: ", ";\n\tfont-size: ", ";\n\tfont-weight: ", ";\n\tfont-style: ", ";\n"])), function (_a) {
    var backgroundColor = _a["data-background-color"];
    return backgroundColor;
}, function (_a) {
    var color = _a["data-color"];
    return color;
}, function (_a) {
    var fontSize = _a["data-font-size"];
    return fontSize;
}, function (_a) {
    var fontWeight = _a["data-font-weight"];
    return fontWeight;
}, function (_a) {
    var fontStyle = _a["data-font-style"];
    return fontStyle;
});
var Style = function (_a) {
    var backgroundColor = _a.backgroundColor, color = _a.color, fontSize = _a.fontSize, fontWeight = _a.fontWeight, italic = _a.italic, children = _a.children;
    var theme = (0, utils_1.useEasyFlexTheme)();
    var processedBackgroundColor = (0, utils_1.useColor)(backgroundColor, undefined);
    var processedColor = (0, utils_1.useColor)(color, undefined);
    var processedFontSize = (0, react_1.useMemo)(function () {
        if (fontSize === undefined) {
            return fontSize;
        }
        var fontSizeValue = (0, utils_1.getFontSize)(theme, fontSize);
        if (theme.fontSizeType === 'rem') {
            return (0, utils_1.toRem)(fontSizeValue);
        }
        return (0, utils_1.toPx)(fontSizeValue);
    }, [fontSize, theme]);
    var processedFontWeight = (0, react_1.useMemo)(function () { return (fontWeight === undefined ? undefined : (0, utils_1.getFontWeight)(theme, fontWeight)); }, [fontWeight, theme]);
    var fontStyle = (0, react_1.useMemo)(function () { return (italic === undefined ? undefined : italic ? 'italic' : 'normal'); }, [italic]);
    return ((0, jsx_runtime_1.jsx)(StyledStyle, __assign({ "data-background-color": processedBackgroundColor, "data-color": processedColor, "data-font-size": processedFontSize, "data-font-weight": processedFontWeight, "data-font-wtyle": fontStyle }, { children: children })));
};
exports.Style = Style;
var templateObject_1;
