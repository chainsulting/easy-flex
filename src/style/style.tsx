import React, { FC, HTMLAttributes, useMemo } from 'react';
import styled from 'styled-components';
import { IColor, IFontSize, IFontStyle, IFontWeight } from '../types';
import { getColor, getFontSize, getFontWeight, toPx, toRem, useEasyFlexTheme } from '../utils';

const StyledStyle = styled.span<{
	'data-background-color'?: string;
	'data-color'?: string;
	'data-font-size'?: string;
	'data-font-weight'?: string | number;
	'data-font-style'?: IFontStyle;
}>`
	display: flex;
	box-sizing: border-box;
	background-color: ${({ 'data-background-color': backgroundColor }) => backgroundColor};
	color: ${({ 'data-color': color }) => color};
	font-size: ${({ 'data-font-size': fontSize }) => fontSize};
	font-weight: ${({ 'data-font-weight': fontWeight }) => fontWeight};
	font-style: ${({ 'data-font-style': fontStyle }) => fontStyle};
`;

export interface IStyleProps extends HTMLAttributes<HTMLSpanElement> {
	backgroundColor?: IColor;
	color?: IColor;
	fontSize?: number | IFontSize;
	fontWeight?: number | IFontWeight;
	italic?: boolean;
}

export const Style: FC<IStyleProps> = ({ backgroundColor, color, fontSize, fontWeight, italic, children }) => {
	const theme = useEasyFlexTheme();

	const processedBackgroundColor = useMemo<string | undefined>(
		() => (backgroundColor === undefined ? undefined : getColor(theme, backgroundColor)),
		[backgroundColor, theme]
	);

	const processedColor = useMemo<string | undefined>(
		() => (color === undefined ? undefined : getColor(theme, color)),
		[color, theme]
	);

	const processedFontSize = useMemo<string | undefined>(() => {
		if (fontSize === undefined) {
			return fontSize;
		}
		const fontSizeValue = getFontSize(theme, fontSize);
		if (theme.fontSizeType === 'rem') {
			return toRem(fontSizeValue);
		}
		return toPx(fontSizeValue);
	}, [fontSize, theme]);

	const processedFontWeight = useMemo<string | number | undefined>(
		() => (fontWeight === undefined ? undefined : getFontWeight(theme, fontWeight)),
		[fontWeight, theme]
	);

	const fontStyle = useMemo<IFontStyle | undefined>(
		() => (italic === undefined ? undefined : italic ? 'italic' : 'normal'),
		[italic]
	);

	return (
		<StyledStyle
			data-background-color={processedBackgroundColor}
			data-color={processedColor}
			data-font-size={processedFontSize}
			data-font-weight={processedFontWeight}
			data-font-wtyle={fontStyle}
		>
			{children}
		</StyledStyle>
	);
};
