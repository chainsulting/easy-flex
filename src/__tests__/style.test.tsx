import { render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';
import { Style } from '../style/Style';

describe('<Style />', () => {
	test('should have the basic css settings', async () => {
		const result = render(<Style />);
		expect(result.baseElement).toMatchSnapshot();
	});
});
