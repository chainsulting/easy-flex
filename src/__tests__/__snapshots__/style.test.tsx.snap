// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Style /> should have the basic css settings 1`] = `
.c0 {
  box-sizing: border-box;
}

<body>
  <div>
    <span
      class="c0"
    />
  </div>
</body>
`;
